blocks/extensions.py

```diff
@@ -0,0 +1,224 @@
+    after_every_epoch : bool
+        If ``True``, :meth:`do` is invoked after every epoch.
+    after_every_iteration : bool
+        If ``True``, :meth:`do` is invoked after every iteration.
+    after_training : bool
+        If ``True``, :meth:`do` is invoked after training.
+
+    """
+    def __init__(self, before_first_epoch=False, after_every_epoch=False,
+                 after_every_iteration=False, after_training=False):
+        self._conditions = []
+        if before_first_epoch:
+            self.add_condition(
+                "before_epoch",
+                predicate=lambda log: log.status.epochs_done == 0)
+        if after_every_epoch:
+            self.add_condition("after_epoch")
+        if after_every_iteration:
+            self.add_condition("after_iteration")
+        if after_training:
+            self.add_condition("after_training")
+
+    def add_condition(self, callback_name, predicate=None, arguments=None):
+        """Adds a condition under which a :meth:`do` is called.
+
+        Parameters
+        ----------
+        callback_name : str
+            The name of the callback in which the method.
+        predicate : function
+            A predicate function the main loop's log as the
+            single parameter and returning ``True`` when the method
+            should be called and ``False`` when should not. If ``None``,
+            an always ``True`` predicate is used.
+        arguments : iterable
+            Additional arguments to be passed to :meth:`do`. They will
+            be concatenated with the ones passed from the main loop
+            (e.g. the batch in case of `after_epoch` callback).
+
+        """
+        if not arguments:
+            arguments = []
+        if not predicate:
+            predicate = lambda log: True
+        self._conditions.append((callback_name, predicate, arguments))
+
+    @abstractmethod
+    def do(self, which_callback, *args):
+        """Does the job of the training extension.
+
+        Parameters
+        ----------
+        which_callback : str
+            The name of the callback in the context of which :meth:`do` is
+            run.
+        *args : tuple
+            The arguments from the main loop concatenated with additional
+            arguments from user.
+
+        """
+        pass
+
+    def execute(self, callback_invoked, *from_main_loop):
+        """Execute methods corresponding to the invoked callback."""
+        for callback_name, predicate, arguments in self._conditions:
+            if (callback_name == callback_invoked
+                    and predicate(self.main_loop.log)):
+                self.do(callback_invoked, *(from_main_loop + tuple(arguments)))
+
+
+class FinishAfter(SimpleExtension):
+    """Finishes the training process when triggered.
+
+    Parameters
+    ----------
+    num_epochs : int
+        If not ``None``, training finish is requested after `num_epochs`
```

Comment:
Technically no, just that it will have to be configured by calling `add_condition`. 

Anyway I think I will move `num_epochs` to the `SimpleExtension` constructor. It might be helpful for other extensions.